src/Adapter/Order/CommandHandler/DeleteProductFromOrderHandler.php

```diff
@@ -174,6 +174,28 @@ private function updateOrderCartRules(Order $order, Cart $cart)
             }
         }
 
+        // Finally add the new cart rules that are not in the Order
+        foreach ($newCartRules as $newCartRule) {
+            foreach ($order->getCartRules() as $orderCartRuleData) {
+                if ($newCartRule['id_cart_rule'] == $orderCartRuleData['id_cart_rule']) {
+                    // This cart rule is already present no need to add it
+                    continue 2;
+                }
+            }
+
+            // Add missing order cart rule
+            $cartRule = new CartRule($newCartRule['id_cart_rule']);
+
+            $orderCartRule = new OrderCartRule();
+            $orderCartRule->id_order = $order->id;
+            $orderCartRule->id_cart_rule = $newCartRule['id_cart_rule'];
+            $orderCartRule->id_order_invoice = $order->getInvoicesCollection()->getLast();
+            $orderCartRule->name = $newCartRule['name'];
+            $orderCartRule->value = Tools::ps_round($cartRule->getContextualValue(true), $computingPrecision);
+            $orderCartRule->value_tax_excl = Tools::ps_round($cartRule->getContextualValue(false), $computingPrecision);
+            $orderCartRule->save();
```

Comment:
I did something similar here https://github.com/PrestaShop/PrestaShop/pull/19646/files#diff-5d60a3a71cc2e386cad2cf1bd26deeb1R430

Reply:
Yeah I was waiting for the QA approval to warn you I didn't plan to do as much but I ended doing almost the same thing 😅 

Although your implementation is better via the service, so the refacto from your PR is meaningful The good point is that I added behat tests for all those use cases (only for delete not cancel) So you can have fun break and refacto everything with peace of mind 👍